ktbridge/plugin/src/main/kotlin/app/cash/quickjs/ktbridge/plugin/BridgedCallRewriter.kt

```diff
@@ -0,0 +1,327 @@
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getPropertyGetter
+import org.jetbrains.kotlin.ir.util.irCall
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+/**
+ * Rewrites calls to `createBridgeToJs` with two arguments:
+ *
+ * ```
+ * val helloService = createBridgeToJs(JsEchoService("hello"), EchoJsAdapter)
+ * ```
+ *
+ * to the overload that takes three arguments:
+ *
+ * ```
+ * val helloService = createBridgeToJs(
+ *   service = JsEchoService("hello"),
+ *   jsAdapter = EchoJsAdapter,
+ *   block = fun (bridgedCall: BridgedCall<JsEchoService>): ByteArray {
+ *     return when {
+ *       bridgedCall.funName == "echo" -> {
+ *         bridgedCall.result(
+ *           bridgedCall.service.echo(
+ *             bridgedCall.parameter()
+ *           )
+ *         )
+ *       }
+ *       else -> bridgedCall.unexpectedFunction()
+ *     }
+ *   }
+ * )
+ * ```
+ *
+ * This synthesizes a function that routes calls by their name, decodes parameters, and encodes the
+ * result.
+ */
+@ObsoleteDescriptorBasedAPI // TODO(jwilson): is there an alternative?
+class BridgedCallRewriter(
+  private val pluginContext: IrPluginContext,
+  private val backingField: IrField,
+  private val initializer: IrExpressionBody,
+  private val initializerCall: IrCall,
+) {
+  private val irFactory: IrFactory
+    get() = pluginContext.irFactory
+
+  private val createBridgeToJsFunction3Arg = pluginContext.referenceFunctions(CREATE_BRIDGE_TO_JS)
+    .single { it.descriptor.valueParameters.size == 3 }
+  private val classBridgedCall = pluginContext.referenceClass(BRIDGED_CALL) ?: error("TODO")
+  private val serviceInterface = initializerCall.getTypeArgument(0) ?: error("TODO")
+  private val bridgedCallOfTType = classBridgedCall.typeWith(serviceInterface)
+
+  private val function: IrSimpleFunction = irFactory.buildFun {
+    name = Name.special("<no name provided>")
+    returnType = pluginContext.symbols.byteArrayType
+  }.apply {
+    parent = backingField
+    addValueParameter(
+      name = "bridgedCall",
+      type = bridgedCallOfTType,
+      origin = IrDeclarationOrigin.DEFINED
+    )
+  }
+
+  fun rewrite() {
+    val irBuilder = DeclarationIrBuilder(
+      generatorContext = pluginContext,
+      symbol = IrSimpleFunctionSymbolImpl(),
+      startOffset = UNDEFINED_OFFSET,
+      endOffset = UNDEFINED_OFFSET
+    )
+
+    val iBridgedFunctionsWhen = irBuilder.irBridgedFunctionsWhen()
+
+    function.body = IrBlockBodyImpl(
+      startOffset = irBuilder.startOffset,
+      endOffset = irBuilder.endOffset,
+    ).apply {
+      statements += irBuilder.irReturn(iBridgedFunctionsWhen)
+    }
+
+    initializer.expression = irCall(initializerCall, createBridgeToJsFunction3Arg).apply {
+      putValueArgument(2, irBuilder.irFunction1Expression())
+    }
+  }
+
+  /**
+   * ```
+   * when {
+   *   bridgedCall.funName == "function1" -> ...
+   *   bridgedCall.funName == "function2" -> ...
```

Comment:
Maybe an opportunity for generating a perfect hash function in the future. I've never found an actual place I could do this and have been wanting to find one for like 5 years.

Reply:
One other option is to implement fancier dispatching so we can have limited skew between declaring interface and implementing interface.

In particular I’d like to be able to introduce new parameters that have default values and have things keep working.